### PR TITLE
Bitstamp: Implement UpdateOrderExecutionLimits and testing

### PR DESCRIPTION
**Pull request by Beadko**
# PR Description

Feature/Bitstamp: OrderExecutionLimits and testing

Fixes # (issue)

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [x] go test ./... -race
- [x] golangci-lint run
- [x] TestUpdateOrderExecutionLimits

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/bitstamp/bitstamp_type_convert.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	t := &struct {
+		*Alias
+		MinimumOrder string `json:"minimum_order"`
+	}{
+		Alias: (*Alias)(p),
+	}
+
+	err := json.Unmarshal(data, t)
+	if err == nil {
+		if m := currRE.FindString(t.MinimumOrder); len(m) > 0 {
```
Comment (shazbert) [inline exchanges/bitstamp/bitstamp_type_convert.go:23]: This is slightly more performant than using regular expression:
```diff
diff --git a/exchanges/bitstamp/bitstamp_type_convert.go b/exchanges/bitstamp/bitstamp_type_convert.go
index abb72525a..959fa670f 100644
--- a/exchanges/bitstamp/bitstamp_type_convert.go
+++ b/exchanges/bitstamp/bitstamp_type_convert.go
@@ -2,12 +2,11 @@ package bitstamp

 import (
        "encoding/json"
-       "regexp"
+       "fmt"
        "strconv"
+       "strings"
 )

-var currRE = regexp.MustCompile(`^[\d.]+`)
-
 // UnmarshalJSON deserializes JSON, and timestamp information.
 func (p *TradingPair) UnmarshalJSON(data []byte) error {
        type Alias TradingPair
@@ -19,11 +18,15 @@ func (p *TradingPair) UnmarshalJSON(data []byte) error {
        }

        err := json.Unmarshal(data, t)
-       if err == nil {
-               if m := currRE.FindString(t.MinimumOrder); len(m) > 0 {
-                       p.MinimumOrder, err = strconv.ParseFloat(m, 64)
-               }
+       if err != nil {
+               return err
+       }
+
+       before, _, found := strings.Cut(t.MinimumOrder, " ")
+       if !found {
+               return fmt.Errorf("unhandled minimum order string: %s", t.MinimumOrder)
        }

+       p.MinimumOrder, err = strconv.ParseFloat(before, 64)
        return err
 }
```

File: exchanges/bitstamp/bitstamp_wrapper.go [hunk 2]
```diff
@@ -315,6 +318,37 @@ func (b *Bitstamp) UpdateTradablePairs(ctx context.Context, forceUpdate bool) er
...
+			continue
+		}
+		pair, err := currency.NewPairFromString(symbols[x].Name)
+		if err != nil {
+			return err
+		}
+		limits = append(limits, order.MinMaxLevel{
+			Asset:                  a,
+			Pair:                   pair,
+			PriceStepIncrementSize: 1 / math.Pow10(info.CounterDecimals),
```
Comment (shazbert) [inline exchanges/bitstamp/bitstamp_wrapper.go:342]: If you can add in amount step that would be good as well:
```diff
diff --git a/exchanges/bitstamp/bitstamp_wrapper.go b/exchanges/bitstamp/bitstamp_wrapper.go
index 314ab47a7..96bd9e2b1 100644
--- a/exchanges/bitstamp/bitstamp_wrapper.go
+++ b/exchanges/bitstamp/bitstamp_wrapper.go
@@ -337,10 +337,11 @@ func (b *Bitstamp) UpdateOrderExecutionLimits(ctx context.Context, a asset.Item)
                        return err
                }
                limits = append(limits, order.MinMaxLevel{
-                       Asset:                  a,
-                       Pair:                   pair,
-                       PriceStepIncrementSize: 1 / math.Pow10(info.CounterDecimals),
-                       MinimumBaseAmount:      info.MinimumOrder,
+                       Asset:                   a,
+                       Pair:                    pair,
+                       PriceStepIncrementSize:  1 / math.Pow10(info.CounterDecimals),
+                       AmountStepIncrementSize: 1 / math.Pow10(info.BaseDecimals),
+                       MinimumBaseAmount:       info.MinimumOrder,
                })
        }
        if err := b.LoadLimits(limits); err != nil {
```

**Review by shazbert — CHANGES_REQUESTED** [review]
Good stuff! just some minor changes. 🚀

File: exchanges/bitstamp/bitstamp_type_convert.go (hunk 1 above)
Comment (Beadko) [reply exchanges/bitstamp/bitstamp_type_convert.go]: That is done

File: exchanges/bitstamp/bitstamp_wrapper.go (hunk 2 above)
Comment (Beadko) [reply exchanges/bitstamp/bitstamp_wrapper.go]: Step amount added

File: exchanges/bitstamp/bitstamp_type_convert.go (hunk 1 above)
Comment (gbjk) [reply exchanges/bitstamp/bitstamp_type_convert.go]: The currency suffix isn't documented, ergo could be subject to change.
If we use Cut like this then I think we should handle t.MinimumOrder as a fallback when not found.

File: exchanges/bitstamp/bitstamp_type_convert.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+	err := json.Unmarshal(data, t)
+	if err != nil {
+		return err
+	}
+	before, _, found := strings.Cut(t.MinimumOrder, " ")
+	if !found {
+		return fmt.Errorf("unhandled minimum order string: %s", t.MinimumOrder)
+	}
+	p.MinimumOrder, err = strconv.ParseFloat(before, 64)
```
Comment (gbjk) [inline exchanges/bitstamp/bitstamp_type_convert.go:28]: The currency suffix isn't documented, ergo could be subject to change.
If we use Cut like this then I think we should handle t.MinimumOrder as a fallback when not found.
```suggestion
	minOrderStr := t.MinimumOrder
	if prefix, _, found := strings.Cut(t.MinimumOrder, " "); found {
		minOrderStr = prefix
	}
	p.MinimumOrder, err = strconv.ParseFloat(minOrderStr, 64)
```

**Review by gbjk — CHANGES_REQUESTED** [review]

File: exchanges/bitstamp/bitstamp_type_convert.go (hunk 3 above)
Comment (gbjk) [reply exchanges/bitstamp/bitstamp_type_convert.go:28]: @shazbert Feedback on this proposal ☝️ please

Comment (shazbert) [reply exchanges/bitstamp/bitstamp_type_convert.go:28]: @gbjk Looks good, nice suggestion.

**Comment by shazbert** [comment]
@Beadko Has some linter issues that need to be rectified. Changes look good!

**Comment by Beadko** [comment]
sorry, forgot that fmt needs r-running, all should be good now

**Review by shazbert — APPROVED** [review]
Thanks Bea, nice work! ACK

File: exchanges/bitstamp/bitstamp_wrapper.go [hunk 4]
```diff
@@ -315,6 +318,38 @@ func (b *Bitstamp) UpdateTradablePairs(ctx context.Context, forceUpdate bool) er
...
+		pair, err := currency.NewPairFromString(symbols[x].Name)
+		if err != nil {
+			return err
+		}
+		limits = append(limits, order.MinMaxLevel{
+			Asset:                   a,
+			Pair:                    pair,
+			PriceStepIncrementSize:  1 / math.Pow10(info.CounterDecimals),
+			AmountStepIncrementSize: 1 / math.Pow10(info.BaseDecimals),
+			MinimumBaseAmount:       info.MinimumOrder,
```
Comment (gloriousCode) [inline exchanges/bitstamp/bitstamp_wrapper.go:344]: Are you absolutely sure that the minimum order size for Bitstamp BTC-USD pair is 20 BTC? Shouldn't this be the minimum _quote_ amount?
image

Comment (Beadko) [reply exchanges/bitstamp/bitstamp_wrapper.go]: changed

**Review by shazbert — APPROVED** [review]
changes tACK! 💃 nice work.

**Review by gloriousCode — APPROVED** [review]
Thanks for your contribution!!

**Review by thrasher- — APPROVED** [review]
Thank you @Beadko !

**Pull request merged**